test/baseVisitors.test.ts

```diff
@@ -0,0 +1,333 @@
+  { type: 'TsArrayType', code: 'type a = string[]', mock: mock.method(baseVisitor, 'TsArrayType'), debug: true },
+  //   TsExpressionWithTypeArguments<S>(n: swc.TsExpressionWithTypeArguments, st: S, cb: Callback<S>) {
+  //     cb(n.expression, st)
+
+  //     if (n.typeArguments) {
+  //       cb(n.typeArguments, st)
+  //     }
+  //   }
+  //   TsInterfaceDeclaration<S>(n: swc.TsInterfaceDeclaration, st: S, cb: Callback<S>) {
+  //     cb(n.id, st)
+  //     cb(n.body, st)
+
+  //     for (const ext of n.extends) {
+  //       cb(ext, st)
+  //     }
+
+  //     if (n.typeParams) {
+  //       cb(n.typeParams, st)
+  //     }
+  //   }
+  //   TsInterfaceBody<S>(n: swc.TsInterfaceBody, st: S, cb: Callback<S>) {
+  //     for (const ele of n.body) {
+  //       cb(ele, st)
+  //     }
+  //   }
+  //   TsKeywordType = ignore
+  //   TsPropertySignature<S>(n: swc.TsPropertySignature, st: S, cb: Callback<S>) {
+  //     for (const param of n.params) {
+  //       cb(param, st)
+  //     }
+
+  //     cb(n.key, st)
+
+  //     if (n.typeAnnotation) {
+  //       cb(n.typeAnnotation, st)
+  //     }
+
+  //     if (n.typeParams) {
+  //       cb(n.typeParams, st)
+  //     }
+  //   }
+  //   TsAsExpression<S>(n: swc.TsAsExpression, st: S, cb: Callback<S>) {
+  //     cb(n.expression, st)
+  //     cb(n.typeAnnotation, st)
+  //   }
+  //   TsCallSignatureDeclaration<S>(n: swc.TsCallSignatureDeclaration, st: S, cb: Callback<S>) {
+  //     for (const param of n.params) {
+  //       cb(param, st)
+  //     }
+  //     if (n.typeParams) {
+  //       cb(n.typeParams, st)
+  //     }
+  //     if (n.typeAnnotation) {
+  //       cb(n.typeAnnotation, st)
+  //     }
+  //   }
+  {
+    type: 'TsExpressionWithTypeArguments',
+    code: 'type foo<T> = Array<T>',
+    mock: mock.method(baseVisitor, 'TsExpressionWithTypeArguments'),
+    skip: true,
+  },
+  {
+    type: 'TsInterfaceDeclaration',
+    code: 'interface Foo {}',
+    mock: mock.method(baseVisitor, 'TsInterfaceDeclaration'),
+  },
+  { type: 'TsInterfaceBody', code: 'interface Foo {}', mock: mock.method(baseVisitor, 'TsInterfaceBody') },
+  { type: 'TsKeywordType', code: 'type foo = string', mock: mock.method(baseVisitor, 'TsKeywordType') },
+  {
+    type: 'TsPropertySignature',
+    code: 'type foo = {a: string}',
+    mock: mock.method(baseVisitor, 'TsPropertySignature'),
+  },
+  { type: 'TsAsExpression', code: 'a as b', mock: mock.method(baseVisitor, 'TsAsExpression') },
+  {
+    type: 'TsCallSignatureDeclaration',
+    code: 'type foo = () => void',
+    mock: mock.method(baseVisitor, 'TsCallSignatureDeclaration'),
+  },
+]
+
+for (const options of tests) {
+  if (options.skip) {
+    continue
+  }
+  test(`should call ${options.type}`, () => {
+    options.mock.mock.resetCalls()
+    expect(given(options)).toHaveBeenCalledTimes(1)
+  })
+}
+
+function given(options: (typeof tests)[number]) {
```

Comment:
This is a nitpick, and I know that function declarations and their body are hoisted to the top of the module scope, but I would prefer that this was not used before it was defined.

I generally enable [`no-use-before-define`](https://typescript-eslint.io/rules/no-use-before-define/). Feel free to add that lint rule.

Reply:
As u want, I usually do like that as a DX good practice, those functions are the least important things in the file, they are "blackboxes" that we do not need to see while opening the file
This way, when I open the file, I directly have my tests

Reply:
I do understand the practicality of keeping the relevant parts first. This is what they refer to as bike shedding :smile: